buildenv/buildenv.nix

```diff
@@ -0,0 +1,278 @@
+    );
+
+  createManifestScript = builtins.toFile "create-manifest-script" (
+    builtins.concatStringsSep "" createManifestChunks
+  );
+
+  # Create manifest package as derivation which invokes above script.
+  manifestPackage = builtins.derivation {
+    name = "manifest";
+    inherit system;
+    builder = "/bin/sh";
+    args = [
+      "-eux"
+      createManifestScript
+    ];
+  };
+
+  # Calculate inputSrcs by noting all storePaths for this system's
+  # packages found in the packages list.
+  inputSrcs = builtins.concatMap (
+    package:
+    if package.system == system then
+      (
+        if (builtins.hasAttr "outputs" package) then
+          (
+            # Important: report storePaths rather than strings because
+            # the updated string context populates `inputSrcs` for the
+            # resulting derivation.
+            let
+              outputsToInstall =
+                if (builtins.hasAttr "outputs_to_install" package) then
+                  (builtins.getAttr "outputs_to_install" package)
+                else if (builtins.hasAttr "outputs-to-install" package) then
+                  (
+                    # XXX kebab-case flake lock bug
+                    builtins.getAttr "outputs-to-install" package
+                  )
+                else
+                  null;
+              filteredOutputs =
+                if (outputsToInstall == null) then
+                  (builtins.attrValues package.outputs)
+                else
+                  (builtins.map (x: builtins.getAttr x package.outputs) outputsToInstall);
+            in
+            map (p: builtins.storePath p) filteredOutputs
+          )
+        else
+          [ ]
+      )
+    else
+      [ ]
+  ) manifestLockData.packages;
+
+in
+# Throw a more meaningful error when lockfile version is < 1.
+assert (builtins.hasAttr "lockfile-version" manifestLockData);
+assert manifestLockData."lockfile-version" != "0";
+builtins.derivation {
+  inherit name;
+  builder = "${floxBuildEnv}/lib/builder.pl";
+  outputs = environmentOutputs;
+
+  # Pull in external attributes and those calculated above.
+  inherit
+    activationScripts
+    inputSrcs
+    manifestPackage
+    system
+    ;
+
+  # If the special attribute __structuredAttrs is set to true, the
+  # other derivation attributes are serialised in JSON format and
+  # made available to the builder via the file .attrs.json in the
+  # builder’s temporary directory. This obviates the need for
+  # passAsFile since JSON files have no size restrictions, unlike
+  # process environments.
+  # https://nix.dev/manual/nix/2.18/language/advanced-attributes#adv-attr-structuredAttrs
+  __structuredAttrs = true;
+
+  # This attribute allows builders access to the references graph of
+  # their inputs. The attribute is a list of inputs in the Nix store
+  # whose references graph the builder needs to know. The value of
+  # this attribute should be a list of pairs `[ name1 path1 name2
```

Comment:
suggestion: consider making builder.pl usable without having to build it.
During dev we want to be able to call buildenv.nix without having to restart the devshell on changes to the builder.
It looks like the only `@...@` substituion in the builder are currently `perl` and `storeDir` both of which we can pass through

Reply:
Well you can bypass the `@perl@` issue by simply calling `perl path/to/builder.pl` from within the dev env, and the `@storeDir@` issue can be addressed by adding these lines within the `isStorePath()` function:
```
    if ($storePath =~ "@.*@") {
        $storePath = "/nix/store";                                                                                                                                                                                                            
    }                                                                                                                                                                                                                                         
```

That said, I would much rather take the approach we do with the activation scripts package which is to detect and rebuild the package when changes occur. This guarantees we're using the exact same store paths both in development and in the end package, and it only takes a second to rebuild. There's no Makefile in `flox/cli` to detect and set up that logic as we did with `pkgdb`, so we'd need a just-expert to muck in on that.

Let me know if you want me to add the `$storePath` default as mentioned above.